File: src/network.cpp
#include "weights_loader.cpp"
#include <list>
#include <cmath>
#include <xtensor/xvectorize.hpp>
#include <vector>
#include <xtensor/xrandom.hpp>
#include "data_loader.cpp"
#include <typeinfo>
#include <xtensor/xsort.hpp>


class network{
    private:

        vector<int> nn_structure = {784,16,16,10};

        int layers ;

        unordered_map<int, xt::xarray<double> > W;
        unordered_map<int, xt::xarray<double> > B;

        unordered_map<int, xt::xarray<double> > H;
        unordered_map<int, xt::xarray<double> > Z;

        unordered_map<int, xt::xarray<double> > delta_W;
        unordered_map<int, xt::xarray<double> > delta_B;

        xt::xarray<double> training_data;
        xt::xarray<double> input_data;
        xt::xarray<double> labels;

        int batch_size = 500;
        int max_iter = 2000;
        double alpha = 1;


    public:
        network(){
            layers = size(nn_structure);
            
        }

        void set_weights_and_biases(unordered_map<int, xt::xarray<double> > biases,unordered_map<int, xt::xarray<double> > weights){
            B = biases;
            W = weights;
        
        }

        void initialise_weights(){
            for(int l =1; l < layers; l++){



                W[l] = xt::random::randn<double>({nn_structure[l], nn_structure[l-1]});
                B[l] = xt::random::randn<double>({nn_structure[l]});


            }
        
            

        }
    
        void initialise_deltas(){
            for(int l =1; l < layers; l++){
                delta_W[l] = xt::zeros<double>({W[l].shape(0), W[l].shape(1)});
                delta_B[l] = xt::zeros<double>({B[l].shape(0)});


            }
            

        }

        void initialise_data(xt::xarray<double> x, xt::xarray<double> y){

            input_data = x;
            labels = y;

        }
    
        static double sigmoid(double z){
            return 1.0 / (1.0 + exp(-z));}

        static double sigmoid_gradient(double z){
            double f = sigmoid(z);
            return f*(1-f) ;
        }

        void feedforward(xt::xarray<double> x){

            H[1] = x;

            xt::xarray<double>  z_l;
            auto vecf = xt::vectorize(&network::sigmoid);

            for(int l =1; l < layers; l++){

                z_l = xt::linalg::dot(W[l], H[l]) + B[l];

                H[l+1] = vecf(z_l);            

                Z[l+1] = z_l;

            }
            H[layers] = H[layers] / xt::sum(H[layers]);


        }

        xt::xarray<double> cost_derivative(xt::xarray<double> y){
            xt::xarray<double> deriv = H[layers] - y;
        
            return deriv;

        }

        void backprop(double y){

            xt::xarray<double> vect_label = vectorised_label(y);

            xt::xarray<double> cost_grad = cost_derivative(vect_label);

            auto vecf = xt::vectorize(&network::sigmoid_gradient);
            
            xt::xarray<double> output_deltas = cost_grad * vecf(Z[layers]);

            unordered_map<int, xt::xarray<double> > deltas;

            deltas[layers] = output_deltas;

            for(int l= layers-1; l >0; l-- ){

                delta_B[l] += deltas[l+1];

                delta_W[l] +=  xt::linalg::dot(xt::view(deltas[l+1], xt::all(), xt::newaxis()),xt::transpose(xt::view(H[l], xt::all(), xt::newaxis())));

                if(l > 1){

                    deltas[l] = xt::linalg::dot(xt::transpose(W[l]),deltas[l+1]) * vecf(Z[l]);
                }

            }

            

        }

        void gradient_descent(){

            for(int l =1; l < layers; l++){

                W[l] +=  -(alpha * 1.0 / batch_size ) * delta_W[l];
                B[l] +=  -(alpha * 1.0 / batch_size ) * delta_B[l];


            }


        }
 
        void train(xt::xarray<double> x, xt::xarray<double> y){

            initialise_data(x,y);
            initialise_weights();
            
            int k;
            xt::xarray<double> image;

            for(int i=0; i< max_iter; i++){

                initialise_deltas();
                
                k = (i*batch_size) % labels.size();
                auto [input_batch, labels_batch] = batch(k);

                for(int j=0; j < batch_size; j++){
                    
                    image = xt::row(input_batch,j);
                    feedforward(image);
                    backprop(labels_batch[j]);

                }
                        
            gradient_descent();

            if( i %10 == 0){
                cout << "Iteration " << i << " ";
                batch_accuracy(input_batch,labels_batch);
            }
            

            }

            evaluate();


        }








        xt::xarray<double> vectorised_label(int y){

            xt::xarray<double> lab = xt::zeros<double>({10});
            lab[y] = 1.0;

            return lab;
        }

        tuple<xt::xarray<double>,xt::xarray<double>>  batch(int k){
            
            xt::xarray<double> batch_labels = xt::zeros<double>({batch_size});
            xt::xarray<double> batch_input = xt::zeros<double>({batch_size,784});
            
            for(int i=0; i < batch_size; i++){

                batch_labels[i] = labels[k+i];

                for(int j=0; j <784; j++){
                    batch_input[i*784 + j]  = input_data[(i+k)*784 + j];

                }

                
            }            


             return {batch_input, batch_labels};
        }

        void batch_accuracy(xt::xarray<double> batch_inputs, xt::xarray<double> batch_labels){

            double accuracy = 0.0;
            xt::xarray<double> image;

            for(int i=0; i < batch_size; i++){
                
                image = xt::row(batch_inputs,i);

                feedforward(image);

                int predicted = xt::argmax(H[layers])[0];

                
                if(predicted == batch_labels[i]){
                    accuracy += 1.0;}

            }

            accuracy /= batch_size;
            accuracy *= 100.0;

            cout << "Accuracy of " << accuracy << "%" << endl;

        }


        void evaluate(){

            double accuracy = 0.0;
            xt::xarray<double> image;

            for(int i=0; i < labels.size(); i++){
                
                image = xt::row(input_data,i);

                feedforward(image);

                int predicted = xt::argmax(H[layers])[0];

                
                if(predicted == labels[i]){
                    accuracy += 1.0;}

            }

            accuracy /= labels.size();
            accuracy *= 100.0;

            cout << "Accuracy of " << accuracy << "%" << endl;

        }





};





int main(){

    unordered_map<int, xt::xarray<double> > biases = biases_loader();

    /*
    unordered_map<int, xt::xarray<double>>::iterator itr;
    for (itr = biases.begin(); 
       itr != biases.end(); itr++) {
    // itr works as a pointer to 
    // pair<string, double> type 
    // itr->first stores the key part and
    // itr->second stores the value part
    cout << itr->first << "  " << 
            itr->second << endl;
    }
    */

    unordered_map<int, xt::xarray<double>> weights =  weights_loader();
    
    network nn;

    auto [input_data, labels]  = JsonLoader();

    //nn.set_weights_and_biases(biases, weights); //comment out initialise_weights in train
    //nn.evaluate();

    nn.train(input_data,labels);

    return 0;
}

File: py/network.py
import json

import numpy as np
from mnist_loader import load_data


class NeuralNet:
    def __init__(self):
        self._nn_structure = [784, 16, 16, 10]

        self._layers = len(self._nn_structure)

        self._weights = {}  # Weights set up as layer: weight matrix, wji
        self._biases = {}  # Biases set up as layer: bias vector

        self._h = {}  # Activations
        self._z = {}

        self._delta_w = {}  # dictionary which has delta wji matrix for each layer
        self._delta_b = {}  # dictionary which has delta b vectors for each layer

        self._batch_size = 1000
        self._training_data = None
        self._input_data = None
        self._labels = None

        self._max_iter = 2000
        self._alpha = 2.5

    def _initialise_data(self, training_data):
        self._training_data = training_data
        self._input_data = training_data[0]
        self._labels = training_data[1]

    def _initialise_weights(self):
        for l in range(1, self._layers):
            # randn used to generate numbers between -0.5 and 0.5 for the weights and biases
            self._weights[l] = np.random.randn(
                self._nn_structure[l], self._nn_structure[l - 1]
            )  # Creates weight matrix in each layer
            self._biases[l] = np.random.randn(
                self._nn_structure[l]
            )  # Creates bias vector in each layer

    def _initialise_deltas(self):
        for l in range(1, self._layers):
            self._delta_w[l] = np.zeros(
                (self._nn_structure[l], self._nn_structure[l - 1])
            )
            self._delta_b[l] = np.zeros(self._nn_structure[l])

    def _give_vectorised_label(self, y):
        lab = np.zeros(10)
        lab[y] = 1.0
        return lab

    def _give_batch(self, k):
        return self._input_data[k : k + self._batch_size], self._labels[
            k : k + self._batch_size
        ]

    def _reshuffle(self):
        randomize = np.arange(len(self._input_data))
        np.random.shuffle(randomize)
        self._input_data = self._input_data[randomize]
        self._labels = self._labels[randomize]

    def _sigmoid(self, x):
        return 1.0 / (1 + np.exp(-x))

    def _sigmoid_gradient(self, x):
        f = self._sigmoid(x)
        return (1.0 - f) * f

    def _feed_forward(self, x):
        self._h[1] = x

        for l in range(1, len(self._nn_structure)):
            z_l = (
                np.matmul(self._weights[l], self._h[l]) + self._biases[l]
            )  # W h + b = z

            self._h[l + 1] = self._sigmoid(z_l)  # f(z) = h

            self._z[l + 1] = z_l

        # self._h[self._layers] = z_l
        # self._softmax_output_layer()

        self._normalise_output_layer()

    def _softmax_output_layer(self):
        """Use instead of _sigmoid activation layer and normalisation"""
        self._h[self._layers] = np.exp(self._h[self._layers])

        factor = sum(self._h[self._layers])
        self._h[self._layers] /= factor

    def _normalise_output_layer(self):
        factor = sum(self._h[self._layers])
        self._h[self._layers] /= factor

    def _cost_function(self, x, y):
        """Cost for a single image"""

        self._feed_forward(x)
        return np.linalg.norm(y - self._h[self._layers])

    def _batch_cost(self, input_batch, labels_batch):
        """Cost of a batch of images"""

        c = 0
        for x, y in zip(input_batch, labels_batch):
            y = self._give_vectorised_label(y)
            c += self._cost_function(x, y)

        c /= len(labels_batch)
        return c

    def _entire_cost(self):
        c = 0
        for x, y in zip(self._input_data, self._labels):
            y = self._give_vectorised_label(y)
            c += self._cost_function(x, y)

        c /= len(self._labels)
        return c

    def _cost_derivative(self, y):
        return self._h[self._layers] - y

    def _backprop(self, y):
        y = self._give_vectorised_label(y)

        cost_grad = self._cost_derivative(y)

        output_deltas = cost_grad * self._sigmoid_gradient(
            self._z[self._layers]
        )  # delta = dC/dh_l * dh_l/dz_l

        deltas = {self._layers: output_deltas}

        for l in range(self._layers - 1, 0, -1):
            # Chain rule derivative dC/db_l = delta
            self._delta_b[l] += deltas[l + 1]

            # Chain rule derivative dC/dW = delta * h_l
            self._delta_w[l] += np.dot(
                deltas[l + 1][:, np.newaxis], np.transpose(self._h[l][:, np.newaxis])
            )

            # Backpropagation for deltas across different layers
            if l > 1:
                deltas[l] = np.dot(
                    np.transpose(self._weights[l]), deltas[l + 1]
                ) * self._sigmoid_gradient(self._z[l])

    def _gradient_descent(self):
        m = self._batch_size

        for l in range(1, self._layers):
            self._weights[l] += -1.0 / m * self._alpha * self._delta_w[l]
            self._biases[l] += -1.0 / m * self._alpha * self._delta_b[l]

    def train(self, training_data):
        self._initialise_data(training_data)
        self._initialise_weights()
        self.load_from_json()

        num_batches = len(input_data) / self._batch_size

        for i in range(self._max_iter):
            if i % num_batches == 0:
                self._reshuffle()

            self._initialise_deltas()

            k = (i * self._batch_size) % len(input_data)

            input_batch, labels_batch = self._give_batch(k)

            for x, y in zip(input_batch, labels_batch):
                self._feed_forward(x)
                self._backprop(y)

            self._gradient_descent()

            c = self._batch_cost(input_batch, labels_batch)

            if i % 10 == 0:
                print(f"Cost {c.round(8)} for iteration: {i}")

    def save_to_json(self):
        """Save weights and biases after training as json file"""

        weights = {key: value.tolist() for key, value in self._weights.items()}
        biases = {key: value.tolist() for key, value in self._biases.items()}

        with open("data/weights.json", "w", encoding="utf-8") as f:
            json.dump(weights, f, ensure_ascii=False, indent=4)

        with open("data/biases.json", "w", encoding="utf-8") as f:
            json.dump(biases, f, ensure_ascii=False, indent=4)

        print("Weights and biases saved as json files!")

    def load_from_json(self):
        """Load weights from json file"""

        with open("data/weights.json") as w:
            weights = json.load(w)

        with open("data/biases.json") as b:
            biases = json.load(b)

        self._weights = {int(key): np.array(value) for key, value in weights.items()}
        self._biases = {int(key): np.array(value) for key, value in biases.items()}

    def evaluate_training(self):
        accuracy = 0

        i = 0
        for x, y in zip(self._input_data, self._labels):
            self._feed_forward(x)
            soft_output = self._h[self._layers]
            prediction = list(soft_output).index(max(soft_output))

            if prediction == y:
                accuracy += 1
            i += 1
        print(f"Accuracy was {100.0 * accuracy / len(self._input_data)} %")
        return accuracy / len(self._input_data)

    def evaluate(self, test_data):
        input_data, labels = test_data[0], test_data[1]

        accuracy = 0
        i = 0
        for x, y in zip(input_data, labels):
            self._feed_forward(x)
            soft_output = self._h[self._layers]
            prediction = list(soft_output).index(max(soft_output))

            if prediction == y:
                accuracy += 1
            i += 1

        print(f"Accuracy was {100.0 * accuracy / len(input_data)} %")
        return accuracy / len(input_data)


if __name__ == "__main__":
    training_data, validation_data, test_data = load_data()

    input_data = test_data[0]
    labels = test_data[1]

    neural_net = NeuralNet()

    neural_net.train(training_data)
    neural_net.evaluate_training()

    neural_net.save_to_json()
